qubes-gui/src/lib.rs

```diff
@@ -1,61 +1,279 @@
+//! on the server, to avoid deadlocks.  Therefore, the client should buffer its
+//! messages and flush them at every opportunity.  This requirement is a
+//! consequence of how difficult asynchronous I/O is in C, and of the desire to
+//! keep the code as simple as possible.  Implementations in other languages, or
+//! which uses proper asynchronous I/O libraries, SHOULD NOT have this
+//! limitation.
+//!
+//! ## Window IDs
+//!
+//! The Qubes OS GUI protocol refers to each surface by a 32-bit unsigned window
+//! ID.  Zero is reserved and means “no window”.  For instance, using zero for a
+//! window’s parent means that the window does not have a parent.  Otherwise,
+//! agents are free to choose any window ID they wish.  In particular, while X11
+//! limits IDs to a maximum of 2²⁹ - 1, the Qubes OS GUI protocol imposes no
+//! such restriction.
+//!
+//! It is a protocol error for an agent to send a message to a window that does
+//! not exist, including a window which it has deleted.  Because of unavoidable
+//! race conditions, agents may recieve events for windows they have already
+//! destroyed.  Such messages MUST be ignored.
+//!
+//! ## Unrecognized messages
+//!
+//! GUI daemons MUST treat messages with an unknown type as a protocol error.
+//! GUI agents MAY log the headers of such messages and MUST otherwise ignore
+//! them.  The bodies of such messages MUST NOT be logged as they may contain
+//! sensitive data.
+//!
+//! ## Shared memory
+//!
+//! The Qubes GUI protocol uses inter-qube shared memory for all images.  This
+//! shared memory is not sanitized in any way whatsoever, and may be modified
+//! by the other side at any time without synchronization.  Therefore, all
+//! access to the shared memory is `unsafe`.  Or rather, it *would* be unsafe,
+//! were it not that no such access is required at all!  This avoids requiring
+//! any form of signal handling, which is both `unsafe` and ugly.
+//!
+//! ## Differences from the reference implementation
+//!
+//! The reference implementation of the GUI protocol considers the GUI daemon
+//! (the server) to be trusted, while the GUI agent is not trusted.  As such,
+//! the GUI agent blindly trusts the GUI daemon, while the GUI daemon must
+//! carefully validate all data from the GUI agent.
+//!
+//! This Rust implementation takes a different view: *Both* the client and server
+//! consider the other to be untrusted, and all messages are strictly validated.
+//! This is necessary to meet Rust safety requirements, and also makes bugs in
+//! the server easier to detect.
+//!
+//! Additionally, the Rust protocol definition is far, *far* better documented,
+//! and explicitly lists each reference to the X11 protocol specification.  A
+//! future release will not depend on the X11 protocol specification at all,
+//! even for documentation.
 
 #![forbid(missing_docs)]
+#![no_std]
+use core;
+use core::num::NonZeroU32;
+use qubes_castable;
```

Comment:
Imports like `use core` and `use qubes_castable` are a no-op. Only paths with more than one component make sense, because they mean you can refer to the item with only the last component of the path.

Reply:
I think that was `extern crate`. Anyway, not really relevant here.